### PR TITLE
♻️ [RUMF-1440] improve feature flag collection implementation

### PR DESCRIPTION
**Pull request by amortemousque**
## Motivation

Extracting feature flag management to a dedicated component.

## Changes
- Remove feature flag logic from trackViews
- Create featureFlagContexts
- Use featureFlagContexts in viewCollection and errorCollection

## Testing

- [x] Local
- [ ] Staging
- [x] Unit
- [ ] End to end

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/rum-core/src/domain/contexts/featureFlagContext.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  })
+
+  lifeCycle.subscribe(LifeCycleEventType.VIEW_CREATED, ({ startClocks }) => {
+    featureFlagContexts.add({}, startClocks.relative)
+  })
+
+  return {
+    findFeatureFlagEvaluations: (startTime?: RelativeTime) => featureFlagContexts.find(startTime),
+    addFeatureFlagEvaluation: (key: string, value: ContextValue) => {
+      if (isExperimentalFeatureEnabled('feature_flags')) {
```
Comment (liywjl) [inline packages/rum-core/src/domain/contexts/featureFlagContext.ts:29]: **💬 suggestion:**
You could check the FF condition first and just return a `noop`  for `findFeatureFlagEvaluations` and `addFeatureFlagEvaluation`. That way you would not need to add any  `lifeCycle.subscribe`

File: packages/rum-core/src/domain/rumEventsCollection/error/errorCollection.ts [hunk 2]
```diff
@@ -105,14 +105,15 @@ function processError(
...
   const inForeground = foregroundContexts.isInForegroundAt(error.startClocks.relative)
   if (inForeground) {
     rawRumEvent.view = { in_foreground: inForeground }
   }
 
-  const viewContext = viewContexts.findView(error.startClocks.relative)
-  if (viewContext && !isEmptyObject(viewContext.featureFlagEvaluations)) {
-    rawRumEvent.feature_flags = viewContext.featureFlagEvaluations
+  const featureFlagContext = featureFlagContexts.findFeatureFlagEvaluations(error.startClocks.relative)
+  if (featureFlagContext && !isEmptyObject(featureFlagContext)) {
```
Comment (liywjl) [inline packages/rum-core/src/domain/rumEventsCollection/error/errorCollection.ts:115]: **💬 suggestion:**
You probably do not need to check `!isEmptyObject(featureFlagContext)`

File: packages/rum-core/src/domain/contexts/featureFlagContext.spec.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      lifeCycle.notify(LifeCycleEventType.VIEW_CREATED, {
+        startClocks: relativeToClocks(10 as RelativeTime),
+      } as ViewCreatedEvent)
+
+      const featureFlagContext = featureFlagContexts.findFeatureFlagEvaluations()!
+      expect(featureFlagContext).toEqual({})
+    })
+
+    it('should return the feature flag context corresponding to the start time', () => {
+      updateExperimentalFeatures(['feature_flags'])
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/rum-core/src/domain/contexts/featureFlagContext.spec.ts:123]: the `resetExperimentalFeatures` is called in the parent describe

File: packages/rum-core/src/domain/rumEventsCollection/error/errorCollection.ts (hunk 2 above)
Comment (amortemousque) [reply packages/rum-core/src/domain/rumEventsCollection/error/errorCollection.ts:115]: The feature flag context with an empty object is created for each `VIEW_CREATED`. If no they are no calls to `addFeatureFlagEvaluation`, we don't want to see the attribute in Datadog. Same logic applied for `customTimings` in `viewCollections`. Wdyt?

File: packages/rum-core/src/domain/contexts/featureFlagContext.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        const currentContext = featureFlagContexts.find()
+        if (currentContext) {
+          // mutate the current context to avoid creating a new context history entry to save memory
+          assign(currentContext, { [key]: value })
```
Comment (bcaudan) [inline packages/rum-core/src/domain/contexts/featureFlagContext.ts]: **❓ question:** ‏I don't get the intent here, why not do directly:
```ts
context[key] = value
```
no test seems to cover that

Comment (bcaudan) [inline packages/rum-core/src/domain/contexts/featureFlagContext.ts]: **❓ question:** ‏should not we `deepClone` the `value` like we do for other APIs allowing customers to pass their own objects?

File: packages/rum-core/src/domain/contexts/featureFlagContext.spec.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+      clock.tick(10)
+      featureFlagContexts.addFeatureFlagEvaluation('feature', 'two')
+
+      lifeCycle.notify(LifeCycleEventType.VIEW_ENDED, {
+        endClocks: relativeToClocks(30 as RelativeTime),
+      } as ViewEndedEvent)
+      lifeCycle.notify(LifeCycleEventType.VIEW_CREATED, {
+        startClocks: relativeToClocks(30 as RelativeTime),
+      } as ViewCreatedEvent)
```
Comment (bcaudan) [inline packages/rum-core/src/domain/contexts/featureFlagContext.spec.ts:148]: **💬 suggestion:** ‏those statements do not seem to be useful for the test

File: packages/rum-core/src/domain/contexts/featureFlagContext.ts (hunk 4 above)
Comment (amortemousque) [reply packages/rum-core/src/domain/contexts/featureFlagContext.ts]: Indeed good point

**Review by bcaudan — APPROVED** [review]

**Pull request merged**